File: examples/javascript/family_b_weak_always_true.js
/**
 * Family B - The check is weak or always true.
 * Codes: C5, C6, C7, C8, C9, C18, JS3, JS15
 *
 * The assertion passes by construction, accepts almost any output, or compares
 * a stringified form instead of the real value.
 *
 * Jest / Vitest idioms, with a little Mocha + Chai variety. Function names
 * (getResult, getUsers, getUser, computeRatio, divide, render) are placeholders.
 */


// --- C5: always-true check -------------------------------------------------

// BAD: both sides are literals - true by construction
test('C5 - literal vs literal', () => {
  expect(true).toBe(true) // C5 - passes regardless of any production code
})

// BAD: a non-empty object literal is always truthy
test('C5 - truthy literal', () => {
  expect({ ok: true }).toBeTruthy() // C5 - the literal is always truthy
})

// BAD: Chai style, equal literals
test('C5 - chai literals', () => {
  expect(1).to.equal(1) // C5
})

// CLEAN: compare the real result against an expected value
test('C5 clean', () => {
  expect(getResult()).toBe(expectedValue)
})


// --- C6: weak check (toBeTruthy / toBeDefined / .length > 0) ----------------

// BAD: toBeTruthy passes for any non-empty list, dict, or object
test('C6 - truthiness', () => {
  const result = getUsers()
  expect(result).toBeTruthy() // C6 - passes for [wrongUser], {}, anything truthy
})

// BAD: length > 0 does not verify the contents
test('C6 - length positive', () => {
  const result = getUsers()
  expect(result.length).toBeGreaterThan(0) // C6 - passes for [null], wrong data
})

// BAD: toBeDefined accepts null, 0, '', false - anything but undefined
test('C6 - defined only', () => {
  expect(getUser(1).role).toBeDefined() // C6 - 'wrong-role' passes too
})

// CLEAN: assert the actual value
test('C6 clean', () => {
  const result = getUsers()
  expect(result).toHaveLength(3)
  expect(result[0].name).toBe('Alice')
})

// CLEAN (not C6): toHaveProperty checks a key exists - a meaningful contract
test('C6 clean - property existence', () => {
  const response = buildResponse()
  expect(response).toHaveProperty('status')
  expect(response).toHaveProperty('data')
})


// --- C7: self-compare (expect(x).toBe(x)) ----------------------------------

// BAD: same reference on both sides - true by reflexivity
test('C7 - self compare', () => {
  const name = getName()
  expect(name).toBe(name) // C7 - always passes
})

// BAD: Chai self-equal
test('C7 - chai self', () => {
  const result = compute()
  expect(result).to.equal(result) // C7
})

// CLEAN: compare against an expected value
test('C7 clean', () => {
  expect(getName()).toBe('Alice')
})

// CLEAN (not C7): two separate calls - tests caching / identity behavior
test('C7 clean - caching', () => {
  expect(loadModule()).toBe(loadModule()) // verifies the loader returns a cached instance
})


// --- C8: exact equality on a float -----------------------------------------

// BAD: floating-point arithmetic makes exact equality unreliable
test('C8 - float exact', () => {
  expect(computeRatio()).toBe(3.14159) // C8 - may fail on rounding (0.1 + 0.2 problem)
})

// CLEAN: toBeCloseTo tolerates floating-point drift
test('C8 clean', () => {
  expect(computeRatio()).toBeCloseTo(3.14159, 5)
})

// CLEAN: 0 and 1 are safe exact sentinels
test('C8 clean - sentinels', () => {
  expect(emptyRatio()).toBe(0)
  expect(fullRatio()).toBe(1)
})


// --- C9: toThrow() with no error type or message ---------------------------

// BAD: bare toThrow passes for any error - including a typo in the test itself
test('C9 - bare toThrow', () => {
  expect(() => divide(a, b)).toThrow() // C9 - any throw passes, even an unrelated bug
})

// CLEAN: assert the error type and message
test('C9 clean', () => {
  expect(() => divide(10, 0)).toThrow(/division by zero/)
})

// CLEAN: specific error class
test('C9 clean - error class', () => {
  expect(() => divide(10, 0)).toThrow(RangeError)
})


// --- C18: stringified equality ---------------------------------------------

// BAD: String(obj) couples to the toString format, not the real values
test('C18 - String() compare', () => {
  const user = getUser(1)
  expect(String(user)).toBe('User(Alice, 30)') // C18 - checks formatting, not identity
})

// BAD: JSON.stringify is brittle - key order and serialization changes break it
test('C18 - JSON.stringify compare', () => {
  expect(JSON.stringify(getConfig())).toBe('{"debug":true,"level":"info"}') // C18
})

// BAD: template literal flattens the value to a string
test('C18 - template literal', () => {
  const ratio = computeRatio()
  expect(`${ratio.toFixed(2)}`).toBe('3.14') // C18 - repr formatting, not the number
})

// CLEAN: compare the structured value with toEqual / attributes
test('C18 clean', () => {
  const user = getUser(1)
  expect(user.name).toBe('Alice')
  expect(user.age).toBe(30)
})

// CLEAN: toEqual does deep value comparison without stringifying
test('C18 clean - toEqual', () => {
  expect(getConfig()).toEqual({ debug: true, level: 'info' })
})


// --- JS3: snapshot is the only assertion -----------------------------------

// BAD: toMatchSnapshot records whatever the component renders today, bugs
// included. The first run sets the baseline as "correct" with no oracle.
test('JS3 - snapshot only', () => {
  const { container } = render(<PriceTag amount={100} />)
  expect(container).toMatchSnapshot() // JS3 - no explicit assertion on the value
})

// CLEAN: snapshot is fine as a supplement, but assert the meaningful value too
test('JS3 clean', () => {
  render(<PriceTag amount={100} />)
  expect(screen.getByTestId('total')).toHaveTextContent('$110.00') // explicit oracle
})


// --- JS15: comparison wrapped in a boolean (expect(a === b).toBe(true)) ----

// BAD: the === is evaluated to a boolean before expect sees it. On failure
// Jest reports "false is not true" with no values, and a typo'd comparison
// (=== vs ==) hides silently.
test('JS15 - boolean-wrapped equality', () => {
  const result = compute()
  expect(result === 42).toBe(true) // JS15 - opaque, weak diagnostics
})

// BAD: negated form
test('JS15 - boolean-wrapped inequality', () => {
  expect(getStatus() !== 'error').toBe(true) // JS15
})

// CLEAN: let the matcher do the comparison so it can report both sides
test('JS15 clean', () => {
  expect(compute()).toBe(42)
})

File: examples/python/family_b_weak_always_true.py
"""
Family B — The check is weak or always true.
Codes: C5, C6, C6b, C7, C8, C9, C11a, C13, C13b, C14, C16, C18, C25, C34, C42, C44

The assertion passes by construction, accepts almost any output, or checks
an implementation detail rather than a meaningful property.
"""
import pytest
import os
from unittest.mock import patch, MagicMock


# ─── C5: always-true check ────────────────────────────────────────────────────

# BAD: assert True is structurally tautological
def test_c5_assert_true():
    assert True             # C5 — always green

# BAD: non-empty tuple is always truthy
def test_c5_tuple():
    assert (item_a, item_b) # C5 — tuple with two elements is always truthy

# BAD: or True short-circuits to True
def test_c5_or_true():
    assert get_result() or True  # C5 — always True regardless of get_result()

# CLEAN
def test_c5_clean():
    assert get_result() == expected_value


# ─── C6: weak check (truthiness, len > 0, fragment search) ───────────────────

# BAD: only checks the result is truthy, not what it contains
def test_c6_truthiness():
    result = get_users()
    assert result           # C6 — passes for any non-empty list, dict, or object

# BAD: len > 0 doesn't verify the contents
def test_c6_len_positive():
    result = get_users()
    assert len(result) > 0  # C6 — passes for [None], [wrong_user], anything non-empty

# BAD: substring in str(obj) checks formatting, not the actual value
def test_c6_substring_in_repr():
    user = get_user(1)
    assert "Alice" in str(user)  # C6 — couples to str() format, not identity

# CLEAN: assert the actual value
def test_c6_clean():
    result = get_users()
    assert len(result) == 3
    assert result[0].name == "Alice"

# CLEAN (not C6): in web/HTTP layer, presence of a response IS the assertion
def test_c6_http_response_clean(client):
    response = client.get("/health")
    assert response           # not C6 — HTTP layer: presence means 200 OK

# CLEAN (not C6): boolean predicates are not weak
def test_c6_isinstance_clean():
    assert isinstance(get_backend(), Backend)

def test_c6_path_predicate_clean(tmp_path):
    p = build_artifact(tmp_path)
    assert p.exists()
    assert p.is_dir()


# ─── C6b: assertion coupled to positional argument via computed index ─────────

# BAD: index computed from another call — fragile and opaque
def test_c6b_computed_index():
    mock_fn = MagicMock()
    do_work(mock_fn)
    idx = expected_args.index("target")
    assert mock_fn.call_args.args[idx] == "target"  # C6b — positional coupling

# CLEAN: assert by name, not position
def test_c6b_clean():
    mock_fn = MagicMock()
    do_work(mock_fn)
    mock_fn.assert_called_once_with(target="target")


# ─── C7: self-comparison ──────────────────────────────────────────────────────

# BAD: both sides are the same name — true by reflexivity
def test_c7_self_compare():
    name = get_name()
    assert name == name     # C7 — always True

# BAD: in unittest style
class TestSelfCompare:
    def test_c7(self):
        result = compute()
        self.assertEqual(result, result)  # C7

# CLEAN: compare against an expected value
def test_c7_clean():
    assert get_name() == "Alice"

# CLEAN (not C7): two separate calls — tests __eq__ or caching behavior
def test_c7_two_calls_clean():
    assert load_module() is load_module()  # tests that the loader is cached

# CLEAN (not C7): test also checks x != peer → testing __eq__ semantics
def test_c7_eq_semantics_clean():
    a = Point(1, 2)
    b = Point(1, 2)
    c = Point(3, 4)
    assert a == b           # __eq__ returns True for same coords
    assert a != c           # __eq__ returns False for different coords


# ─── C8: exact equality on float ─────────────────────────────────────────────

# BAD: floating-point arithmetic makes exact equality unreliable
def test_c8_float_eq():
    assert compute_ratio() == 3.14159  # C8 — may fail on rounding

# CLEAN
def test_c8_approx_clean():
    assert compute_ratio() == pytest.approx(3.14159, rel=1e-6)

# CLEAN: 0.0 and 1.0 are safe sentinels
def test_c8_sentinel_clean():
    assert empty_ratio() == 0.0
    assert full_ratio() == 1.0


# ─── C9: pytest.raises too broad ─────────────────────────────────────────────

# BAD: any exception passes — including a typo in the test itself
def test_c9_broad_raises():
    with pytest.raises(Exception):   # C9 — too broad
        divide(a, b)

# BAD: no match= parameter
def test_c9_no_match():
    with pytest.raises(ValueError):  # C9 LOW — type is specific but message unchecked
        parse(bad_input)

# CLEAN: specific type + message pattern
def test_c9_clean():
    with pytest.raises(ZeroDivisionError, match="division by zero"):
        divide(10, 0)


# ─── C11a: self-confirming literal ───────────────────────────────────────────

# BAD: assigns a value then asserts the same value — only tests Python assignment
def test_c11a_self_confirming():
    product.price = 100
    assert product.price == 100  # C11a — tests Python attribute assignment

# CLEAN: test the derived computation, not the assigned value
def test_c11a_clean():
    product.price = 100
    assert get_price_with_tax(product) == 110  # 100 + 10% tax


# ─── C13: mock assertion misspelled or not called ────────────────────────────

# BAD: missing parentheses — returns a bound method, doesn't run the check
def test_c13_missing_parens():
    mock_fn = MagicMock()
    do_work(mock_fn)
    mock_fn.assert_called_once   # C13 — no (), nothing checked

# BAD: invented method name (not a real mock assertion)
def test_c13_invented_method():
    mock_fn = MagicMock()
    do_work(mock_fn)
    mock_fn.assert_called_twice()     # C13 — not a real method, silently no-ops

# BAD: assert_not_called_once doesn't exist
def test_c13_wrong_name():
    mock_fn = MagicMock()
    mock_fn.assert_not_called_once    # C13 — invented

# CLEAN
def test_c13_clean():
    mock_fn = MagicMock()
    do_work(mock_fn)
    mock_fn.assert_called_once_with(expected_arg)


# ─── C13b: patch without autospec ────────────────────────────────────────────

# BAD: typos in argument names silently pass
@patch("mymodule.send_email")
def test_c13b_no_autospec(mock_send):
    send_notification(user)
    mock_send.assert_called_once_with(to=user.email, subjct="Hi")  # C13b — typo undetected

# CLEAN: autospec enforces the real signature
@patch("mymodule.send_email", autospec=True)
def test_c13b_autospec_clean(mock_send):
    send_notification(user)
    mock_send.assert_called_once_with(to=user.email, subject="Hi")


# ─── C14: golden file written from actual output ─────────────────────────────

# BAD: first run records whatever the code produces — including bugs
def test_c14_golden_from_output(tmp_path):
    golden = tmp_path / "expected.txt"
    actual = render(template)
    if not golden.exists():
        golden.write_text(actual)   # C14 — records today's output as truth
    assert golden.read_text() == actual

# CLEAN: golden file is committed with a known-correct value, not generated
def test_c14_committed_golden(tmp_path, golden_dir):
    expected = (golden_dir / "expected.txt").read_text()
    assert render(template) == expected


# ─── C16: depends on time, randomness, or sleep ──────────────────────────────

# BAD: datetime.now() is not frozen
def test_c16_raw_datetime():
    import datetime
    assert is_expired(datetime.datetime.now(), ttl=0) is False  # C16 — clock not frozen

# BAD: random without seed
def test_c16_random_without_seed():
    import random
    result = random.choice(candidates)
    assert result in candidates  # C16 — passes but provides no real coverage

# BAD: time.sleep makes tests flaky under CI load
def test_c16_sleep():
    import time
    start_task()
    time.sleep(0.1)             # C16 — fragile wall-clock wait
    assert task_done()

# CLEAN: clock frozen with freezegun
def test_c16_frozen_clock():
    import datetime
    from freezegun import freeze_time
    with freeze_time("2024-01-01"):
        assert is_expired(datetime.datetime.now(), ttl=3600) is False

# CLEAN: random with seed
def test_c16_seeded_random():
    import random
    random.seed(42)
    result = random.choice(candidates)
    assert result == expected_with_seed_42


# ─── C18: string/repr comparison ─────────────────────────────────────────────

# BAD: couples to str() format — changes when __str__ changes
def test_c18_str_compare():
    user = get_user(1)
    assert str(user) == "User(Alice, 30)"  # C18 — checks formatting, not values

# BAD: f-string comparison
def test_c18_fstring_compare():
    assert f"{result:.2f}" == "3.14"       # C18 — repr formatting, not the float

# CLEAN: compare attributes directly
def test_c18_clean():
    user = get_user(1)
    assert user.name == "Alice"
    assert user.age == 30


# ─── C25: xfail without strict=True ──────────────────────────────────────────

# BAD: if the test unexpectedly passes, pytest reports XPASS but doesn't fail
@pytest.mark.xfail
def test_c25_no_strict():
    assert buggy_function() == expected  # C25 — XPASS silently accepted

# CLEAN
@pytest.mark.xfail(strict=True, reason="known bug #42")
def test_c25_strict_clean():
    assert buggy_function() == expected


# ─── C34: suboptimal assertion form ──────────────────────────────────────────

# BAD: each has a clearer idiomatic alternative
def test_c34_not_in():
    assert not "x" in items         # C34 — use: assert "x" not in items

def test_c34_len_zero():
    assert len(result) == 0         # C34 — use: assert not result

def test_c34_eq_true():
    assert is_valid() == True       # C34 — use: assert is_valid()

def test_c34_eq_false():
    assert is_valid() == False      # C34 — use: assert not is_valid()

def test_c34_eq_none():
    assert get_result() == None     # C34 — use: assert get_result() is None

def test_c34_ne_none():
    assert get_result() != None     # C34 — use: assert get_result() is not None

# CLEAN
def test_c34_clean():
    assert "x" not in items
    assert not result
    assert is_valid()
    assert not is_valid()
    assert get_result() is None
    assert get_result() is not None


# ─── C42: assert on a generator expression / lambda (always truthy) ──────────

# BAD: the generator object is truthy regardless of its contents
def test_c42_genexpr():
    assert (x for x in get_items())   # C42 — always passes, even if get_items() is empty

# BAD: a lambda object is always truthy
def test_c42_lambda():
    assert lambda: compute()          # C42 — never calls compute(), always truthy

# CLEAN: materialize and compare
def test_c42_clean():
    items = list(get_items())
    assert items == [1, 2, 3]


# ─── C44: numeric tautology (len()/abs() is never negative) ──────────────────

# BAD: len() is never negative, so this is always true
def test_c44_len_ge_zero():
    assert len(get_items()) >= 0      # C44 — passes for any input

# BAD: abs() >= 0 is always true
def test_c44_abs():
    assert abs(compute()) >= 0        # C44

# CLEAN: compare to a real expected count
def test_c44_clean():
    assert len(get_items()) == 3
